### DISCUSSION TITLE
【2026腾讯犀牛鸟】Qwen3-ASR ncnn 移植与多平台部署

### DISCUSSION THREAD

**Discussion by AiChiTuDouPian** [Tencent/ncnn#6849] [Show and tell]
# Qwen3-ASR-ncnn 纯 C++ 端到端语音识别技术报告

> 版本：V1.1
> 项目路径：https://github.com/AiChiTuDouPian/qwen3-ASR-ncnn
> 目标：将 Qwen/Qwen3-ASR（0.6B）通过 pnnx 转为 ncnn，并用纯 C++ 实现「音频文件 → 文本」的端到端推理，输出与 PyTorch 原版**逐字一致**；最小第三方依赖，Windows / Linux 均可编译运行。
> 新增优化：decode放在VULKAN上加速，lm_head和audio tower加上部分优化！目前性能瓶颈集中在audio tower部分，在pnnx针对动态图的适配问题，当前使用CPU计算，速度偏慢，prefill和decode速度相比于cpu版本提升5.x倍。
---

## 1. 项目概述

| 项 | 内容 |
|----|------|
| 模型 | Qwen3-ASR（语音识别，0.6B 变体） |
| 文本解码器 | hidden=1024，28 层，GQA，KV-cache 增量解码 |
| 音频塔 | d_model=896，18 层 encoder，输出经 proj → 1024 注入文本侧 |
| 推理后端 | ncnn（`build-ncnn-omp`：AVX2 + OpenMP 已开启） |
| 依赖 | ncnn + dr_mp3.h（单文件，公有领域，MP3 解码）；**不依赖 ffmpeg** |
| 端到端一致性 | test1.mp3 与 PyTorch 原版 `model.generate` 转录**逐字一致** |
| 多语言 | 中文 / 日语实测通过，解码与输出对语言通用（UTF-8） |

- 直接输入音频地址（`.mp3` / `.wav` / …）即输出文本
- 支持可变音频长度输入。

---

## 2. 整体架构

端到端数据流分三大块：

```
音频文件(.mp3/.wav)
   │  [1] 音频前端（纯 C++ 无外部依赖）
   ▼
Whisper Mel [T, 128]  (16kHz, STFT n_fft=400 hop=160, 128 mel)
   │  [2] 音频塔（Audio Tower，ncnn 子图）
   ▼
音频 embeddings [L, 1024]   (L ≈ T/ 下采样)
   │  在 <|audio_pad|> 位置注入文本 decoder
   ▼
[3] 文本解码器（LLM，ncnn 子图，28 层，贪心生成）
   ▼
token 序列 → byte-level BPE 解码 → UTF-8 文本
```

- **[1] 音频前端**：解封装 → 重采样 → mel 提取，全部 C++ 手写，无外部音频库。
1. MP3 解码用 dr_mp3.h，将多声道取平均转单声道
输出：单声道 float PCM，范围 [-1, 1]，附带原始采样率
WAV 文件则是手写 RIFF 解析器，支持 8/16/24/32-bit PCM 和 IEEE float。
2. 重采样到 16kHz
Whisper mel 前端要求 16kHz 输入。用的是 Kaiser 窗 sinc 插值，这里使用128个Mel Register
比如 44.1kHz → 16kHz，ratio ≈ 0.363，每个输出样本对周围 128 个输入样本做加权求和。
3. Mel 频谱图提取
n_fft = 400，hop_length = 160
center=True，reflect padding（首尾各补 200 样本）
Hann 窗（从 hann_window.bin 加载）
输出 201 个频率 bin 的 power spectrum（幅度平方）
DFT 实现： 纯 C++ DFT，计算 twiddle factor 矩阵，201×400 = 80400 个值对，每帧 400 样本乘 Hann 窗后逐 bin 计算。
4. Mel 滤波器组：
128 个 mel 滤波器，从 mel_filters.bin 加载 [128×201]
矩阵乘：mel[128, T] = filters[128,201] × power[201, T]
5.Log-mel 归一化：
最终输出 [T, 128]（时间×频率），T = 帧数。
- **[2] Audio Tower**：把 mel 编码成与文本同维度的音频表征。
1. MP3 解码用 dr_mp3.h，将多声道取平均转单声道
2. 旋转位置编码
3. transformer encoder
- **[3] 文本解码器**：标准因果 LLM 解码，音频表征在 `<|audio_pad|>` 位置「注入」。
1. prefill
2. GQA Decoder
3. 贪心解码

---

### 3.1 音频前端（纯 C++ 自包含）

| 步骤 | 实现 | 说明 |
|------|------|------|
| MP3 解码 | `dr_mp3.h` v0.7.4（mackron/dr_libs，单文件公有领域） | 无需 ffmpeg |
| WAV 解码 | 原生 PCM 读取器 | 支持 8/16/24/32-bit int + 32-bit float，多声道→单声道 |
| 重采样 | Kaiser 窗 sinc 重采样 | 目标 16kHz，L=64，β=12 |
| Mel | Whisper 式 | center=True(reflect) STFT：n_fft=400，hop=160，Hann 窗 → 功率谱 → `mel_filters[128,201]` 投影 → `log10(clamp 1e-10)` → `clamp(global_max-8)` → `(x+4)/4` |

> 权重 `hann_window.bin`(400) 与 `mel_filters.bin`(128×201) 由 `python/export_mel_weights.py` 从 HuggingFace `WhisperFeatureExtractor` **精确导出**（mel_filters 需转置为 `[mel, freq]` 行主序）。

### 3.2 音频塔（Audio Tower）

- 输入 mel `[T, 128]` → **3× Conv2d**（480 通道，kernel=2？实际 stride=2、pad=1、GELU 逐级下采样）→ `conv_out`（7680 → 896）。
- 逐 chunk **Sinusoids 位置编码**（分母 447）。
- **18 层 encoder**：dim=896，heads=14，head_dim=64，attn scale=1/8。
- `ln_post` → `proj1`（GELU）→ `proj2`（→ 1024，对齐文本维度）。

> **注意力实现要点**：HuggingFace 源码虽写了滑动窗口，但 `sdpa` 实际不消费 `cu_seq_lens`，故为**全局双向注意力**。C++ 实现采用 `window = total`（全序列），与官方行为一致。

### 3.3 文本解码器（LLM Decoder，28 层）

单层结构（与标准 LLaMA/Qwen 同构）：

```
x → input RMSNorm → QKV 线性
Q,K → q_norm / k_norm → RoPE（rotate_half）
Q,K,V → GQA 注意力（KV-cache 增量 O(T)）
→ o_proj → +残差
→ post RMSNorm → MLP(SiLU) → +残差
```

- `HIDDEN = 1024`，KV-cache 随序列长度增量增长，避免重复计算。
- `lm_head` 与 `embed` 权重共享（tie）。

### 3.4 音频注入机制

HuggingFace 用 `masked_scatter(audio_mask, audio_features)` 在 `<|audio_pad|>` 位置填入音频表征。C++ 对应 `process_one(audio_emb)`：在 prompt 中每个 `<|audio_pad|>` 位置，用音频塔输出替换该位置的 embedding 后再送入解码器。

### 3.5 提示词构造（Chat Template，语言中立）

```cpp
// PREFIX:  <|im_start|> system \n <|im_end|> \n <|im_start|> user \n <|audio_start|>
const std::vector<int> PREFIX_IDS = {
    IM_START_ID, 8948/*system*/, 198/*\n*/, IM_END_ID, 198,
    IM_START_ID, 872/*user*/, 198, AUDIO_START_ID };

// SUFFIX:  <|audio_end|> <|im_end|> \n <|im_start|> assistant \n
const std::vector<int> SUFFIX_IDS = {
    AUDIO_END_ID, IM_END_ID, 198, IM_START_ID, 77091/*assistant*/, 198 };
```

完整 prompt = `PREFIX + L×AUDIO_PAD_ID + SUFFIX`。**该模板不含任何语言强制指令**，语言由模型按音频自动检测。

### 3.6 特殊 Token 与解码

| 名称 | id | 含义 |
|------|----|------|
| `IM_START_ID` | 151644 | `<|im_start|>` |
| `IM_END_ID` | 151645 | `<|im_end|>`（生成停止符） |
| `AUDIO_START_ID` | 151669 | `<|audio_start|>` |
| `AUDIO_END_ID` | 151670 | `<|audio_end|>` |
| `ASR_TEXT_ID` | 151704 | `<asr_text>`（正文开始标记） |
| `AUDIO_PAD_ID` | 151676 | `<|audio_pad|>`（音频注入位） |

- 词表 `vocab.txt`：151643 行（id 0..151642），`SPECIAL_BEGIN=151643` 以上为特殊 token。
- `decode()` 跳过 `id >= 词表大小` 的特殊 token，故 `<|im_end|>` / `<asr_text>` 等不会出现在正文。
- **Tokenizer 为 byte-level BPE（decode-only）**：`vocab[id]` 是 byte-level unicode 串 → 按 `BYTE_ENCODER` 反查回原始字节 → `UTF-8 decode`。该逻辑对**所有语言通用**（中文/日文/英文共用同一条路径）。

---

### 4.1 逐模块数值对齐（与 PyTorch 参考）

| 模块 | 比对方式 | 结果 |
|------|----------|------|
| 音频塔 | C++ `run_audio_tower` vs PyTorch 参考 | cos = **1.0** |
| 文本解码器（28 层） | `python/golden_layers.py` 逐层 plain-RoPE 对比 `debug_dump_layers` | 每层 cos = **0.999996 ~ 1.0** |
| 端到端生成 | `python/golden_generate.py` 逐算子贪心 vs C++ `generate()` | 输出 token 序列逐位一致（首步 top1 logit 30.03 vs 30.03） |

> 关键坑（已排除）：早期用 `txt(input_ids=...)` 生成的黄金参考走的是 **MRoPE**，而 C++ 与逐算子参考用 **plain rotate_half RoPE**，两者本不对齐；改用 safetensors 直读 + plain-RoPE 的 `golden_layers.py` 后，逐层余弦达到 1.0。

### 4.2 端到端文本一致性

| 音频 | C++ 输出 | PyTorch 原版 | 一致性 |
|------|----------|--------------|--------|
| test1.mp3（中文，~9s） | 其形也，天若惊鸿，宛若游龙，荣耀秋菊，华茂春松，仿佛。 | 逐字相同 | ✅ 完全一致 |
| test2.mp3（中文，~122s） | 《洛神赋》片段（含模型真实吐出的 ` Chinese` 语言标记） | — | 结果正确 |
| japantest.mp3（日语） | 何も僕の願いが叶うなら、君と同じものが欲しいんだ… | — | ✅ 结果正确 |

### 4.3 音频前端 Mel 对齐

C++ 提取的 mel 与官方 `WhisperFeatureExtractor` 特征 **cos = 0.998**，差异仅来自 `dr_mp3` / `librosa` 解码与重采样实现的微小差别。

---

### 5.1 多语言验证

日语 `japantest.mp3` 实测输出纯净日文，**无 ` Chinese` 前缀**——证明 ` Chinese` 仅为模型对中文音频**自生成的语言标记**，对日语音频不会强加中文。解码与输出修复对所有 UTF-8 语言通用。

> **中文验证**  

[洛神赋.mp3](https://github.com/user-attachments/files/30375547/default.mp3)

 **结果 **： Chinese其形也，天若惊鸿，晚若游龙，荣耀秋菊，华茂春松。仿佛曦若青云之碧月，飘摇曦若流风之回雪。远而望之，皎若太阳升朝霞；或而察之，浊若浮渠出露波。农先得中，修短合度；坚若薛城，腰如约素。严谨秀像，好至成路，方则无家，千华浮玉。云髻峨峨，修眉连娟，丹唇外朗，皓齿内鲜。明眸善睐，夜伏成全，龟兹艳逸，一镜体闲。柔情绰态，媚于语言，其服旷世，古相映图。披罗衣之璀灿兮，而摇臂之华裾。待金翠之首饰，缀明珠以耀躯。见远游之文侣，夜雾消之清居。微幽兰之芳霭兮，不持出于山隅。于是乎烟纵体，以遨以兮。左以彩毛，右应贵奇。攘号万于神虎兮，彩湍濑之玄之。于情悦其淑美兮，心振荡而不。无良媒以接欢兮，托微波而通辞。愿成宿之先达兮，借玉佩以邀之。皆佳人之信修兮，锵其礼而凝诗。抗穷壁以鹤于兮，止前渊而为妻。执卷卷之款石兮，聚四灵之我兮。干交甫之弃言兮，畅游浴而胡疑。

> **英语验证**  
[en.mp3](https://github.com/user-attachments/files/30383130/en.mp3)
 **结果 **：Is this what it feels like to be lonely? And though I'm not afraid, no, it won't be the same without you.
```log
[audio] decoded sr=44100 samples=925103 (20.98s)
[audio] resampled to 16k: 335638 samples (20.98s)
[asr_main] mel T=2097 (from audio)
[audio_tower] layer 1/18
[audio_tower] layer 4/18
[audio_tower] layer 7/18
[audio_tower] layer 10/18
[audio_tower] layer 13/18
[audio_tower] layer 16/18
[audio_tower] layer 18/18
[asr_main] audio frames L=273
[prefill] 0/288
[prefill] 256/288
[prefill] done: 24555 ms (288 tokens)
  [step] +151.4 ms (next_tok=6364)
  [step] +149.7 ms (next_tok=151704)
Is  [step] +151.9 ms (next_tok=419)
 this  [step] +151.6 ms (next_tok=1128)
 what  [step] +153.1 ms (next_tok=432)
 it  [step] +152.7 ms (next_tok=11074)
 feels  [step] +150.8 ms (next_tok=1075)
 like  [step] +150.0 ms (next_tok=311)
 to  [step] +151.3 ms (next_tok=387)
 be  [step] +150.4 ms (next_tok=39566)
 lonely  [step] +153.8 ms (next_tok=30)
?  [step] +152.2 ms (next_tok=1597)
 And  [step] +151.2 ms (next_tok=3498)
 though  [step] +152.8 ms (next_tok=358)
 I  [step] +151.6 ms (next_tok=2776)
'm  [step] +154.4 ms (next_tok=537)
 not  [step] +151.8 ms (next_tok=16575)
 afraid  [step] +152.2 ms (next_tok=11)
,  [step] +151.0 ms (next_tok=902)
 no  [step] +150.8 ms (next_tok=11)
,  [step] +151.8 ms (next_tok=432)
 it  [step] +151.6 ms (next_tok=2765)
 won  [step] +151.3 ms (next_tok=944)
't  [step] +151.8 ms (next_tok=387)
 be  [step] +151.1 ms (next_tok=279)
 the  [step] +152.8 ms (next_tok=1852)
 same  [step] +152.7 ms (next_tok=2041)
 without  [step] +152.7 ms (next_tok=498)
 you  [step] +154.2 ms (next_tok=13)
.  [step] +155.0 ms (next_tok=151645)
```
> **日语验证**  

[心做.mp3](https://github.com/user-attachments/files/30375549/default.mp3)

 **结果 **：何も僕の願いが叶うなら、君と同じものが欲しいんだ。でも僕には存在しないから、じゃあせめてここに。

```log
[prefill] 0/262
[prefill] 256/262
[prefill] done: 12924 ms (262 tokens)
  [step] +79.7 ms (next_tok=10769)
  [step] +81.3 ms (next_tok=151704)
  [step] +74.2 ms (next_tok=58899)
  [step] +75.1 ms (next_tok=243)
  [step] +77.7 ms (next_tok=15767)
  [step] +74.3 ms (next_tok=103227)
  [step] +76.5 ms (next_tok=16586)
  [step] +79.7 ms (next_tok=28195)
  [step] +75.9 ms (next_tok=100027)
  [step] +74.1 ms (next_tok=29197)
  [step] +74.4 ms (next_tok=125028)
  [step] +75.7 ms (next_tok=5373)
  [step] +74.4 ms (next_tok=101956)
  [step] +75.6 ms (next_tok=133281)
  [step] +76.3 ms (next_tok=134265)
  [step] +80.9 ms (next_tok=131540)
  [step] +74.1 ms (next_tok=125486)
  [step] +75.5 ms (next_tok=1773)
  [step] +74.7 ms (next_tok=71561)
  [step] +77.1 ms (next_tok=58899)
  [step] +75.8 ms (next_tok=243)
  [step] +77.6 ms (next_tok=125009)
  [step] +76.1 ms (next_tok=47606)
  [step] +79.7 ms (next_tok=129306)
  [step] +75.6 ms (next_tok=53931)
  [step] +74.8 ms (next_tok=5373)
  [step] +75.7 ms (next_tok=125817)
  [step] +77.2 ms (next_tok=29491)
  [step] +75.5 ms (next_tok=71242)
  [step] +77.8 ms (next_tok=126646)
  [step] +75.7 ms (next_tok=125695)
  [step] +79.8 ms (next_tok=19655)
  [step] +75.9 ms (next_tok=1773)
  [step] +81.3 ms (next_tok=151645)
[asr_main] generated 32 tokens
何も僕の願いが叶うなら、君と同じものが欲しいんだ。でも僕には存在しないから、じゃあせめてここに。

```

## 6. 结论

Qwen3-ASR-ncnn 已实现**纯 C++、零 ffmpeg 依赖**的「音频文件 → 文本」端到端推理：

- **数值正确性**：音频塔 cos=1.0、文本解码器 28 层逐层 cos≈1.0、端到端 test1.mp3 与 PyTorch 原版逐字一致。
- **自包含前端**：MP3/WAV 解码 + Kaiser 重采样 + Whisper mel 全部 C++ 手写，mel cos=0.998。
- **多语言**：中文/日语实测通过，UTF-8 解码与输出修复对所有语言通用。
- **性能**：CPU 上 decode ~200ms/token；主要瓶颈在 prefill 串行，batch prefill 为下一步关键优化。

 **项目地址** ：https://github.com/AiChiTuDouPian/qwen3-ASR-ncnn
 **模型地址 **：https://huggingface.co/xxzigou/wen3-ASR-ncnn
